### DISCUSSION TITLE
Using PIXI.js to solve performance problem (shaders?) for reaction-diffusion program

### DISCUSSION THREAD

**Discussion by butchern** [pixijs/pixijs#7705] [Q&A]
Hello,
I am doing an A level programming project, where I am building a gray-scott reaction-diffusion system. (see here: https://www.karlsims.com/rd.html)

I implemented a prototype using regular JavaScript, with the help of a Coding train video (https://www.youtube.com/watch?v=BV9ny785UNc). It became clear I would need to use the GPU to have it run at a bearable speed. (Here is my prototype using P5 on the CPU: https://reaction-diffusion.20butchernathan.repl.co/ - press space to start, it logs the framerate to the console). Even without any customisation it is incredibly slow.

I found a version using shaders with THREE.js on the GPU, which ran at a very high speed, made by pmnelia (http://pmneila.github.io/jsexp/grayscott/). However THREE.js is 3d renderer, and I want to use PIXI, as it is better suited for making this 2D program.

I was working through Kittykatattack’s “learn PIXI” book to learn general usage of PIXI, but it doesn’t have much that is relevant to my project, for example there is no mention of shaders.

In P5 I could use the “pixels array” to change the colour of every pixel, I haven’t found a way to do this in PIXI. I was considering using a series of rectangles instead but I don’t know if that is practical in PIXI, and shaders appear to be a much more efficient option. 

However I don’t know how to (or even if I can) use shaders in PIXI to do this, are there any good resources I should use, or do you have any other advice?

Thank you in advance!

**Comment by ivanpopelyshev** [comment]
https://pixijs.io/examples/#/mesh-and-shaders/shadertoy-mesh.js

>  are there any good resources I should use,

issues in this repo , threads in https://www.html5gamedevs.com/forum/13-frameworks/

there were many of question like "how to do "pixels array"" or "how to make a shader/filter/whatever". The thing is, no one actually compiled  answers into articles :) wanna be a hero?

===

regarding your question, i dont have enough time right now to look in what is actually you need, so you have to wait for someone else or dig in. In general, if you have 10 steps you want to do, and you know how to make 2 of them in webgl - you can use pixi to speed-up the process and make all 10 fast. However if you dont know how webgl works, you'll have same problems in pure webgl, in threejs and in pixi.

**Reply by butchern** [reply]
Thank you for your response, I'll have a look into the shadertoy and into pure weblgl. 
This was my first post on Github so I apologise for not doing it in the right place!

**Comment by butchern** [comment]
Hello,
Thanks for your suggestion of the forum.
I was able to use a Youtube tutorial [https://www.youtube.com/watch?v=wIC-CqsUplw](url) to gain a good understanding of how the shaders work and how to use them with Pixi, alongside a website containing documentation for the language the shaders use and also the shaders used in Pmnelia’s code which are written in this language the shaders use (GLSL) and do essentially what I need. These mean that I am confident I can program my own shaders to perform the main process my program needs; however, I have come across another problem.
The reaction diffusion system is a form of cellular automata – it defines the next generation of cells via the previous ones (in this case the new concentrations of chemicals in (and hence the colour of) each pixel, from the previous concentrations (colours)). In Pixi, the shader is used as a filter, applied over the top of a texture. Shaders cannot have their own values as inputs, so to make my system work I have to alternate a pair of textures, one of which has the shader applied, and is displayed, and then the other texture uses the previous texture’s values for the same process, and so on.
My problem is that I don’t know if there is a way to effectively “read” the previous iteration of the shader. As the shader is used as a filter, the texture remains unchanged, so I don’t know how to read the colour values (concentrations) of the previous texture, to define the next instance of the shader.
Am I missing an obvious way of doing this?
Furthermore, it would be ideal if I could create my pair of textures within pixi’s program to apply my shader to instead of having to import a pair of images that are the size I want, is there any way to do this?
I’m more than happy to share what I’ve learned and what I’ve done with it. Thanks.

**Reply by ivanpopelyshev** [reply]
https://github.com/pixijs/pixijs/wiki/v5-Hacks#filters

still works: https://www.pixiplayground.com/#/edit/vb2GGM0S1m8_Cp0v6jmH4

**Reply by ivanpopelyshev** [reply]
or you can do all that with mesh-shader and a number of renderTextures, that way you wont have to read sources of FilterSystem to understand how it works

**Reply by butchern** [reply]
Thanks very much for the continued advice. 
I have spent the days since your reply searching to try and understand  render textures / mesh shaders and the hack you showed as well, but I haven't been able to find any resources that have given me a good understanding of the methods that would enable me to go away and properly use them.
Are you aware of any tutorials on either of these approaches, or someone who can help me implement them?

**Reply by ivanpopelyshev** [reply]
Sorry, cant work on that input, its too ambigious. Make a demo that doesnt work - I can fix it, and it'll be another thread like previous ones. There's no WebGL renderer that has complete articles on those things, its not only pixi problem.

**Reply by ivanpopelyshev** [reply]
Basically, someone will have to remove all other projects and do full-time on shader demos and making plugins like https://www.youtube.com/channel/UC7M-Wz4zK8oikt6ATcoTwBA does. Until then, there will be only my threads on how i help people in my free time.
